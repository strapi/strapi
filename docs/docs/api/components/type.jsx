import React from 'react';

export default function Type({ children }) {
  return (
    <span
      style={{
        color: '#017501',
      }}
    >
      &lt;{children}&gt;
    </span>
  );
}
