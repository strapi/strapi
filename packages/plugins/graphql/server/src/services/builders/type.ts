import { isArray, isString, isUndefined, constant } from 'lodash/fp';
import { nonNull, list, objectType } from 'nexus';
import { contentTypes } from '@strapi/utils';
import type { Struct } from '@strapi/types';

import type { Context } from '../types';

export type TypeBuildersOptions = {
  builder: any;
  attributeName: string;
  attribute: any;
  contentType: any;
  context: Context;
};

export default (context: Context) => {
  const { strapi } = context;
  const getGraphQLService = strapi.plugin('graphql').service;

  const extension = getGraphQLService('extension');

  /**
   * Add a scalar attribute to the type definition
   *
   * The attribute is added based on a simple association between a Strapi
   * type and a GraphQL type (the map is defined in `strapiTypeToGraphQLScalar`)
   */
  const addScalarAttribute = (options: TypeBuildersOptions) => {
    const { builder, attributeName, attribute } = options;

    const { mappers } = getGraphQLService('utils');

    const gqlType = mappers.strapiScalarToGraphQLScalar(attribute.type);

    builder.field(attributeName, { type: gqlType });
  };

  /**
   * Add a component attribute to the type definition
   *
   * The attribute is added by fetching the component's type
   * name and using it as the attribute's type
   */
  const addComponentAttribute = (options: TypeBuildersOptions) => {
    const { builder, attributeName, contentType, attribute } = options;

    let localBuilder = builder;

    const { naming } = getGraphQLService('utils');
    const { getContentTypeArgs } = getGraphQLService('builders').utils;
    const { buildComponentResolver } = getGraphQLService('builders').get('content-api');

    const type = naming.getComponentNameFromAttribute(attribute);

    if (attribute.repeatable) {
      localBuilder = localBuilder.list;
    }

    const targetComponent = strapi.getModel(attribute.component);

    const resolve = buildComponentResolver({
      contentTypeUID: contentType.uid,
      attributeName,
      strapi,
    });

    const args = getContentTypeArgs(targetComponent, {
      multiple: !!attribute.repeatable,
      isNested: true,
    });

    localBuilder.field(attributeName, { type, resolve, args });
  };

  /**
   * Add a dynamic zone attribute to the type definition
   *
   * The attribute is added by fetching the dynamic zone's
   * type name and using it as the attribute's type
   */
  const addDynamicZoneAttribute = (options: TypeBuildersOptions) => {
    const { builder, attributeName, contentType } = options;

    const { naming } = getGraphQLService('utils');
    const { ERROR_CODES } = getGraphQLService('constants');
    const { buildDynamicZoneResolver } = getGraphQLService('builders').get('content-api');

    const { components } = contentType.attributes[attributeName];

    const isEmpty = components.length === 0;
    const type = naming.getDynamicZoneName(contentType, attributeName);

    const resolve = isEmpty
      ? // If the dynamic zone don't have any component, then return an error payload
        constant({
          code: ERROR_CODES.emptyDynamicZone,
          message: `This dynamic zone don't have any component attached to it`,
        })
      : //  Else, return a classic dynamic-zone resolver
        buildDynamicZoneResolver({
          contentTypeUID: contentType.uid,
          attributeName,
        });

    builder.list.field(attributeName, { type, resolve });
  };

  /**
   * Add an enum attribute to the type definition
   *
   * The attribute is added by fetching the enum's type
   * name and using it as the attribute's type
   */
  const addEnumAttribute = (options: TypeBuildersOptions) => {
    const { builder, attributeName, contentType } = options;

    const { naming } = getGraphQLService('utils');

    const type = naming.getEnumName(contentType, attributeName);

    builder.field(attributeName, { type });
  };

  /**
   * Add a media attribute to the type definition
   */
  const addMediaAttribute = (options: TypeBuildersOptions) => {
    const { naming } = getGraphQLService('utils');
    const { getContentTypeArgs } = getGraphQLService('builders').utils;
    const { buildAssociationResolver } = getGraphQLService('builders').get('content-api');
    const extension = getGraphQLService('extension');

    const { builder } = options;
    const { attributeName, attribute, contentType } = options;
    const fileUID = 'plugin::upload.file';

    if (extension.shadowCRUD(fileUID).isDisabled()) {
      return;
    }

    const fileContentType = strapi.contentTypes[fileUID];

    const resolve = buildAssociationResolver({
      contentTypeUID: contentType.uid,
      attributeName,
      strapi,
    });

    const args = attribute.multiple
      ? getContentTypeArgs(fileContentType, { isNested: true })
      : undefined;

    const typeName = naming.getTypeName(fileContentType);

    if (attribute.multiple) {
      builder.field(`${attributeName}_connection`, {
        type: naming.getRelationResponseCollectionName(fileContentType),
        resolve,
        args,
      });

      builder.field(attributeName, {
        type: nonNull(list(typeName)),
        async resolve(...args: unknown[]) {
          const res = await resolve(...args);
          return res.nodes ?? [];
        },
        args,
      });
    } else {
      builder.field(attributeName, {
        type: typeName,
        async resolve(...args: unknown[]) {
          const res = await resolve(...args);
          return res.value;
        },
        args,
      });
    }
  };

  /**
   * Add a polymorphic relational attribute to the type definition
   */
  const addPolymorphicRelationalAttribute = (options: TypeBuildersOptions) => {
    const { GENERIC_MORPH_TYPENAME } = getGraphQLService('constants');
    const { naming } = getGraphQLService('utils');
    const { buildAssociationResolver } = getGraphQLService('builders').get('content-api');

    let { builder } = options;
    const { attributeName, attribute, contentType } = options;

    const { target } = attribute;
    const isToManyRelation = attribute.relation.endsWith('Many');

    if (isToManyRelation) {
      builder = builder.list;
    }
    // todo[v4]: How to handle polymorphic relation w/ entity response collection types?
    //  -> Currently return raw polymorphic entities

    const resolve = buildAssociationResolver({
      contentTypeUID: contentType.uid,
      attributeName,
      strapi,
    });

    // If there is no specific target specified, then use the GenericMorph type
    if (isUndefined(target)) {
      builder.field(attributeName, {
        type: GENERIC_MORPH_TYPENAME,
        resolve,
      });
    }

    // If the target is an array of string, resolve the associated morph type and use it
    else if (isArray(target) && target.every(isString)) {
      const type = naming.getMorphRelationTypeName(contentType, attributeName);

      builder.field(attributeName, { type, resolve });
    }
  };

  /**
   * Add a regular relational attribute to the type definition
   */
  const addRegularRelationalAttribute = (options: TypeBuildersOptions) => {
    const { naming } = getGraphQLService('utils');
    const { getContentTypeArgs } = getGraphQLService('builders').utils;
    const { buildAssociationResolver } = getGraphQLService('builders').get('content-api');
    const extension = getGraphQLService('extension');

    const { builder } = options;
    const { attributeName, attribute, contentType } = options;

    if (extension.shadowCRUD(attribute.target).isDisabled()) {
      return;
    }

    const isToManyRelation = attribute.relation.endsWith('Many');

    const resolve = buildAssociationResolver({
      contentTypeUID: contentType.uid,
      attributeName,
      strapi,
    });

    const targetContentType = strapi.getModel(attribute.target);

    const typeName = naming.getTypeName(targetContentType);

    const args = isToManyRelation
      ? getContentTypeArgs(targetContentType, { isNested: true })
      : undefined;

    const resolverScope = `${targetContentType.uid}.find`;
    const resolverPath = `${naming.getTypeName(contentType)}.${attributeName}`;

    extension.use({ resolversConfig: { [resolverPath]: { auth: { scope: [resolverScope] } } } });

    if (isToManyRelation) {
      builder.field(`${attributeName}_connection`, {
        type: naming.getRelationResponseCollectionName(targetContentType),
        resolve,
        args,
      });

      extension.use({
        resolversConfig: { [`${resolverPath}_connection`]: { auth: { scope: [resolverScope] } } },
      });

      builder.field(attributeName, {
        type: nonNull(list(typeName)),
        async resolve(...args: unknown[]) {
          const res = await resolve(...args);
          return res.nodes ?? [];
        },
        args,
      });
    } else {
      builder.field(attributeName, {
        type: typeName,
        async resolve(...args: unknown[]) {
          const res = await resolve(...args);
          return res.value;
        },
        args,
      });
    }
  };

  const isNotPrivate = (contentType: Struct.Schema) => (attributeName: string) => {
    return !contentTypes.isPrivateAttribute(contentType, attributeName);
  };

  const isNotDisabled = (contentType: Struct.Schema) => (attributeName: string) => {
    return extension.shadowCRUD(contentType.uid).field(attributeName).hasOutputEnabled();
  };

  return {
    /**
     * Create a type definition for a given content type
     * @param contentType - The content type used to created the definition
     * @return {NexusObjectTypeDef}
     */
    buildTypeDefinition(contentType: Struct.Schema) {
      const utils = getGraphQLService('utils');

      const { getComponentName, getTypeName } = utils.naming;
      const {
        isStrapiScalar,
        isComponent,
        isDynamicZone,
        isEnumeration,
        isMedia,
        isMorphRelation,
        isRelation,
      } = utils.attributes;

      const { attributes, modelType } = contentType;

      const attributesKey = Object.keys(attributes);

      const name = (modelType === 'component' ? getComponentName : getTypeName).call(
        null,
        contentType
      );

      return objectType({
        name,
        definition(t) {
          // add back the old id attribute on contentType if v4 compat is enabled
          if (
            modelType !== 'component' &&
            isNotDisabled(contentType)('id') &&
            strapi.plugin('graphql').config('v4CompatibilityMode', false)
          ) {
            t.nonNull.id('id', {
              deprecation: 'Use `documentId` instead',
            });
          }

          if (modelType === 'component' && isNotDisabled(contentType)('id')) {
            t.nonNull.id('id');
          }

          if (modelType !== 'component' && isNotDisabled(contentType)('documentId')) {
            t.nonNull.id('documentId');
          }

          if (strapi.plugin('graphql').config('v4CompatibilityMode', false)) {
            t.nonNull.field('attributes', {
              deprecation: 'Use root level fields instead',
              type: name,
              resolve: (parent) => parent,
            });

            t.nonNull.field('data', {
              deprecation: 'Use root level fields instead',
              type: name,
              resolve: (parent) => parent,
            });
          }

          /** Attributes
           *
           * Attributes can be of 7 different kind:
           * - Scalar
           * - Component
           * - Dynamic Zone
           * - Enum
           * - Media
           * - Polymorphic Relations
           * - Regular Relations
           *
           * Here, we iterate over each non-private attribute
           * and add it to the type definition based on its type
           */
          attributesKey
            // Ignore private attributes
            .filter(isNotPrivate(contentType))
            // Ignore disabled fields (from extension service)
            .filter(isNotDisabled(contentType))
            // Add each attribute to the type definition
            .forEach((attributeName) => {
              const attribute = attributes[attributeName];

              // We create a copy of the builder (t) to apply custom
              // rules only on the current attribute (eg: nonNull, list, ...)
              let builder: any = t;

              if (attribute.required) {
                builder = builder.nonNull;
              }

              const options: TypeBuildersOptions = {
                builder,
                attributeName,
                attribute,
                contentType,
                context,
              };

              // Enums
              if (isEnumeration(attribute)) {
                addEnumAttribute(options);
              }

              // Scalars
              else if (isStrapiScalar(attribute)) {
                addScalarAttribute(options);
              }

              // Components
              else if (isComponent(attribute)) {
                addComponentAttribute(options);
              }

              // Dynamic Zones
              else if (isDynamicZone(attribute)) {
                addDynamicZoneAttribute(options);
              }

              // Media
              else if (isMedia(attribute)) {
                addMediaAttribute(options);
              }

              // Polymorphic Relations
              else if (isMorphRelation(attribute)) {
                addPolymorphicRelationalAttribute(options);
              }

              // Regular Relations
              else if (isRelation(attribute)) {
                addRegularRelationalAttribute(options);
              }
            });
        },
      });
    },
  };
};
