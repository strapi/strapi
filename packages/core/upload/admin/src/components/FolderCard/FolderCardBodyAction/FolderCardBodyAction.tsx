// TODO: find a better naming convention for the file that was an index file before
import { Box, BoxProps } from '@strapi/design-system';
import { NavLink, LinkProps } from 'react-router-dom';
import { styled } from 'styled-components';

const BoxOutline = styled(Box)`
  &:focus {
    outline: 2px solid ${({ theme }) => theme.colors.primary600};
    outline-offset: -2px;
  }
`;

const BoxTextDecoration = styled(BoxOutline)`
  text-decoration: none;
`;

interface FolderCardBodyActionProps extends BoxProps {
  to?: LinkProps['to'];
}

export const FolderCardBodyAction = ({ to, ...props }: FolderCardBodyActionProps) => {
  if (to) {
    return (
      <BoxTextDecoration
        // padding needed to give outline space to appear
        // since FolderCardBody needs overflow hidden property
        padding={1}
        tag={NavLink}
        maxWidth="100%"
        to={to}
        {...props}
      />
    );
  }

  return <BoxOutline padding={1} tag="button" type="button" maxWidth="100%" {...props} />;
};
