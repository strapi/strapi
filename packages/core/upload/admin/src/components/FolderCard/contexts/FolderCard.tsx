import { createContext, useContext } from 'react';

export const FolderCardContext = createContext<{
  id?: string;
}>({});

export function useFolderCard() {
  return useContext(FolderCardContext);
}
