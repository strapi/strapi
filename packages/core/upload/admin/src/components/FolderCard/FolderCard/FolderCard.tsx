import * as React from 'react';

import { Box, CardAction, Flex, BoxProps } from '@strapi/design-system';
import { Folder } from '@strapi/icons';
import { NavLink, LinkProps } from 'react-router-dom';
import { styled } from 'styled-components';

import { FolderCardContext } from '../contexts/FolderCard';

const FauxClickWrapper = styled.button`
  height: 100%;
  left: 0;
  position: absolute;
  opacity: 0;
  top: 0;
  width: 100%;

  &:hover,
  &:focus {
    text-decoration: none;
  }
`;

const StyledFolder = styled(Folder)`
  path {
    fill: currentColor;
  }
`;

const CardActionDisplay = styled(Box)`
  display: none;
`;

const Card = styled(Box)`
  &:hover,
  &:focus-within {
    ${CardActionDisplay} {
      display: ${({ $isCardActions }) => ($isCardActions ? 'block' : '')};
    }
  }
`;

export interface FolderCardProps extends Omit<BoxProps, 'id'> {
  ariaLabel: string;
  children: React.ReactNode;
  id?: string;
  startAction?: React.ReactNode;
  cardActions?: React.ReactNode;
  onClick?: () => void;
  to?: LinkProps['to'];
}

export const FolderCard = React.forwardRef(
  (
    {
      children,
      startAction = null,
      cardActions = null,
      ariaLabel,
      onClick,
      to,
      ...props
    }: FolderCardProps,
    ref
  ) => {
    const generatedId = React.useId();
    const fodlerCtxValue = React.useMemo(() => ({ id: generatedId }), [generatedId]);

    return (
      <FolderCardContext.Provider value={fodlerCtxValue}>
        <Card position="relative" tabIndex={0} $isCardActions={!!cardActions} ref={ref} {...props}>
          <FauxClickWrapper
            to={to || undefined}
            as={to ? NavLink : 'button'}
            type={to ? undefined : 'button'}
            onClick={onClick}
            tabIndex={-1}
            aria-label={ariaLabel}
            aria-hidden
          />

          <Flex
            hasRadius
            borderStyle="solid"
            borderWidth="1px"
            borderColor="neutral150"
            background="neutral0"
            shadow="tableShadow"
            padding={3}
            gap={2}
            cursor="pointer"
          >
            {startAction}

            <Box
              hasRadius
              background="secondary100"
              color="secondary500"
              paddingBottom={2}
              paddingLeft={3}
              paddingRight={3}
              paddingTop={2}
            >
              <StyledFolder width="2.4rem" height="2.4rem" />
            </Box>

            {children}

            <CardActionDisplay>
              <CardAction right={4} position="end">
                {cardActions}
              </CardAction>
            </CardActionDisplay>
          </Flex>
        </Card>
      </FolderCardContext.Provider>
    );
  }
);
