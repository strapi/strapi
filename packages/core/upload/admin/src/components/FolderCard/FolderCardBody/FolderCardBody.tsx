import { Flex, FlexProps } from '@strapi/design-system';
import { styled } from 'styled-components';

import { useFolderCard } from '../contexts/FolderCard';

const StyledBox = styled(Flex)`
  user-select: none;
`;

export const FolderCardBody = (props: FlexProps) => {
  const { id } = useFolderCard();

  return (
    <StyledBox
      {...props}
      id={`${id}-title`}
      data-testid={`${id}-title`}
      alignItems="flex-start"
      direction="column"
      maxWidth="100%"
      overflow="hidden"
      position="relative"
    />
  );
};
