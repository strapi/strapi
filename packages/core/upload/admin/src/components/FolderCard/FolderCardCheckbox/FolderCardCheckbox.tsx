import { Checkbox, Box, CheckboxProps } from '@strapi/design-system';

import { useFolderCard } from '../contexts/FolderCard';

export const FolderCardCheckbox = (props: CheckboxProps) => {
  const { id } = useFolderCard();

  return (
    <Box position="relative" zIndex={2}>
      <Checkbox aria-labelledby={`${id}-title`} {...props} />
    </Box>
  );
};
