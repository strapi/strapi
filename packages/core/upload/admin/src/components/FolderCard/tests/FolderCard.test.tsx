import { Flex, DesignSystemProvider, Typography } from '@strapi/design-system';
import { render } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { MemoryRouter, LinkProps } from 'react-router-dom';

import { FolderCard } from '../FolderCard/FolderCard';
import { FolderCardBody } from '../FolderCardBody/FolderCardBody';
import { FolderCardBodyAction } from '../FolderCardBodyAction/FolderCardBodyAction';
import { FolderCardCheckbox } from '../FolderCardCheckbox/FolderCardCheckbox';

import type { FolderCardProps } from '../FolderCard/FolderCard';

const ID_FIXTURE = 'folder';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useId: () => `${ID_FIXTURE}-3`,
}));

const ComponentFixture = ({
  to,
  ...props
}: {
  to?: LinkProps['to'];
  props?: Partial<FolderCardProps>;
}) => {
  return (
    <DesignSystemProvider>
      <MemoryRouter>
        <FolderCard
          id={ID_FIXTURE}
          ariaLabel="Folder 1"
          startAction={null}
          onClick={() => {}}
          to={to}
          {...props}
        >
          <FolderCardBody>
            <FolderCardBodyAction onClick={() => {}} to={to}>
              <Flex direction="column" alignItems="flex-start">
                <Typography variant="omega" fontWeight="semiBold">
                  Pictures
                </Typography>
              </Flex>
            </FolderCardBodyAction>
          </FolderCardBody>
        </FolderCard>
      </MemoryRouter>
    </DesignSystemProvider>
  );
};

const setup = (props?: Partial<FolderCardProps>) => render(<ComponentFixture {...props} />);

describe('FolderCard', () => {
  test('renders and matches the snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  test('properly calls the onClick callback', async () => {
    const callback = jest.fn();
    const { getByLabelText } = setup({ onClick: callback });

    const button = getByLabelText('Folder 1');
    await userEvent.click(button);

    expect(callback).toHaveBeenCalledTimes(1);
  });

  test('has all required ids set when rendering a start action', () => {
    const { container, getByTestId, getByRole } = setup({
      startAction: <FolderCardCheckbox value={'false'} />,
    });

    expect(container).toMatchSnapshot();
    expect(getByTestId(`${ID_FIXTURE}-3-title`)).toBeInTheDocument();
    expect(getByRole('checkbox', { checked: false })).toBeInTheDocument();
  });

  test('renders as a link with to prop', () => {
    const { getByRole } = setup({
      to: '/michka-page',
    });

    const link = getByRole('link');
    expect(link).toHaveAttribute('href', '/michka-page');
  });

  test('renders as a button without to prop', () => {
    const { getByRole } = setup();

    const button = getByRole('button');
    expect(button).toBeInTheDocument();
  });
});
