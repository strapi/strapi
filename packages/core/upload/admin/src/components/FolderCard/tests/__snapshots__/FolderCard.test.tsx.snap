// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderCard has all required ids set when rendering a start action 1`] = `
.c0 {
  position: relative;
}

.c2 {
  padding-block-start: 12px;
  padding-inline-end: 12px;
  padding-block-end: 12px;
  padding-inline-start: 12px;
  background: #ffffff;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  border-color: #eaeaef;
  box-shadow: 0px 1px 4px rgba(33, 33, 52, 0.1);
  cursor: pointer;
}

.c4 {
  z-index: 2;
  position: relative;
}

.c7 {
  padding-block-start: 8px;
  padding-block-end: 8px;
  padding-inline-start: 12px;
  padding-inline-end: 12px;
  color: #66b7f1;
  background: #eaf5ff;
  border-radius: 4px;
}

.c9 {
  max-width: 100%;
  position: relative;
  overflow: hidden;
}

.c12 {
  padding-block-start: 4px;
  padding-inline-end: 4px;
  padding-block-end: 4px;
  padding-inline-start: 4px;
  max-width: 100%;
}

.c16 {
  right: 16px;
}

.c3 {
  gap: 8px;
  align-items: center;
  flex-direction: row;
  display: flex;
}

.c10 {
  align-items: flex-start;
  flex-direction: column;
  display: flex;
}

.c14 {
  font-size: 1.4rem;
  line-height: 1.43;
  color: currentcolor;
  font-weight: 500;
}

.c18 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c17 {
  position: absolute;
  top: 12px;
  right: 12px;
}

.c5 {
  background: #ffffff;
  width: 2rem;
  height: 2rem;
  border-radius: 4px;
  border: 1px solid #c0c0cf;
  position: relative;
  z-index: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  flex: 0 0 2rem;
}

.c5[data-state='checked'],
.c5[data-state='indeterminate'] {
  border: 1px solid #4945ff;
  background-color: #4945ff;
}

.c5[data-disabled] {
  background-color: #dcdce4;
}

.c5::before {
  content: '';
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 100%;
  height: 100%;
  z-index: -1;
  min-width: 44px;
  min-height: 44px;
}

.c6 {
  display: inline-flex;
  pointer-events: auto!important;
  width: 100%;
  height: 100%;
  cursor: pointer;
  justify-content: center;
  align-items: center;
}

.c1 {
  height: 100%;
  left: 0;
  position: absolute;
  opacity: 0;
  top: 0;
  width: 100%;
}

.c1:hover,
.c1:focus {
  text-decoration: none;
}

.c8 path {
  fill: currentColor;
}

.c15 {
  display: none;
}

.c11 {
  user-select: none;
}

.c13:focus {
  outline: 2px solid #4945ff;
  outline-offset: -2px;
}

<div>
  <div
    class="c0 "
    id="folder"
    tabindex="0"
  >
    <button
      aria-hidden="true"
      aria-label="Folder 1"
      class="c1"
      tabindex="-1"
      type="button"
    />
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <button
          aria-checked="false"
          aria-labelledby="folder-3-title"
          class="c5"
          data-state="unchecked"
          role="checkbox"
          type="button"
          value="false"
        >
          <span
            class="c6"
            data-state="unchecked"
            style="pointer-events: none;"
          />
        </button>
      </div>
      <div
        class="c7"
      >
        <svg
          class="c8"
          fill="currentColor"
          height="2.4rem"
          viewBox="0 0 32 32"
          width="2.4rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M27 9H16.414L13 5.586A1.98 1.98 0 0 0 11.586 5H5a2 2 0 0 0-2 2v18.078A1.926 1.926 0 0 0 4.924 27H27.11A1.89 1.89 0 0 0 29 25.111V11a2 2 0 0 0-2-2M5 7h6.586l2 2H5z"
          />
        </svg>
      </div>
      <div
        class="c9 c10 c11"
        data-testid="folder-3-title"
        id="folder-3-title"
      >
        <button
          class="c12 c13"
          type="button"
        >
          <div
            class="c10"
          >
            <span
              class="c14"
            >
              Pictures
            </span>
          </div>
        </button>
      </div>
      <div
        class="c15"
      >
        <div
          class="c16 c3 c17"
        />
      </div>
    </div>
  </div>
  <span
    class="c18"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;

exports[`FolderCard renders and matches the snapshot 1`] = `
.c0 {
  position: relative;
}

.c2 {
  padding-block-start: 12px;
  padding-inline-end: 12px;
  padding-block-end: 12px;
  padding-inline-start: 12px;
  background: #ffffff;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  border-color: #eaeaef;
  box-shadow: 0px 1px 4px rgba(33, 33, 52, 0.1);
  cursor: pointer;
}

.c4 {
  padding-block-start: 8px;
  padding-block-end: 8px;
  padding-inline-start: 12px;
  padding-inline-end: 12px;
  color: #66b7f1;
  background: #eaf5ff;
  border-radius: 4px;
}

.c6 {
  max-width: 100%;
  position: relative;
  overflow: hidden;
}

.c9 {
  padding-block-start: 4px;
  padding-inline-end: 4px;
  padding-block-end: 4px;
  padding-inline-start: 4px;
  max-width: 100%;
}

.c13 {
  right: 16px;
}

.c3 {
  gap: 8px;
  align-items: center;
  flex-direction: row;
  display: flex;
}

.c7 {
  align-items: flex-start;
  flex-direction: column;
  display: flex;
}

.c11 {
  font-size: 1.4rem;
  line-height: 1.43;
  color: currentcolor;
  font-weight: 500;
}

.c15 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c14 {
  position: absolute;
  top: 12px;
  right: 12px;
}

.c1 {
  height: 100%;
  left: 0;
  position: absolute;
  opacity: 0;
  top: 0;
  width: 100%;
}

.c1:hover,
.c1:focus {
  text-decoration: none;
}

.c5 path {
  fill: currentColor;
}

.c12 {
  display: none;
}

.c8 {
  user-select: none;
}

.c10:focus {
  outline: 2px solid #4945ff;
  outline-offset: -2px;
}

<div>
  <div
    class="c0 "
    id="folder"
    tabindex="0"
  >
    <button
      aria-hidden="true"
      aria-label="Folder 1"
      class="c1"
      tabindex="-1"
      type="button"
    />
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <svg
          class="c5"
          fill="currentColor"
          height="2.4rem"
          viewBox="0 0 32 32"
          width="2.4rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M27 9H16.414L13 5.586A1.98 1.98 0 0 0 11.586 5H5a2 2 0 0 0-2 2v18.078A1.926 1.926 0 0 0 4.924 27H27.11A1.89 1.89 0 0 0 29 25.111V11a2 2 0 0 0-2-2M5 7h6.586l2 2H5z"
          />
        </svg>
      </div>
      <div
        class="c6 c7 c8"
        data-testid="folder-3-title"
        id="folder-3-title"
      >
        <button
          class="c9 c10"
          type="button"
        >
          <div
            class="c7"
          >
            <span
              class="c11"
            >
              Pictures
            </span>
          </div>
        </button>
      </div>
      <div
        class="c12"
      >
        <div
          class="c13 c3 c14"
        />
      </div>
    </div>
  </div>
  <span
    class="c15"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;
